File: minesweeper.cpp
#include <iostream>
#include <cstdlib>
#include <sstream>
#include <time.h>
#include <math.h>
using namespace std;

const int boardSize = 9;//Game Board size minimum 9 (9x9 game board)
const int bombs = int(((boardSize)*(boardSize))/10);//Bombs must be present in 10% of the game tiles
int remainingFlags = bombs;
int remainingTiles = (boardSize*boardSize) - bombs;

//Game Tiles
string coveredTile = "■";
string emptyTile = "□";
string flaggedTile = "🏴";
string bombTile = "💣";

int gameBoard[boardSize][boardSize];
string display[boardSize][boardSize];
int bombCoords[bombs][2];
int score;
enum gameOptions{
    OPEN_TILE = 1,
    FLAG_TILE,
    EXIT_GAME
};

enum tileStates{
    BOMB = -1,
    EMPTY_TILE
};

void printDisplay(){
  //Prints the display board

  for(int a = 0; a <= boardSize -1; a++){
      for(int b = 0; b <= boardSize -1; b++){
        if (b == 0) {
          cout << "               "<< a <<" ▶  " << display[a][b] << "   ";
        }else {
          cout << display[a][b] << "   ";
        }
      }
      cout << endl<< endl;
  }
}


int showOptions(){
    int option;
    score = round((100 - remainingTiles/float(((boardSize*boardSize) - bombs))*100));
    cout << "                   #Minesweeper#" << endl;
    cout << endl;
    cout << endl;
    cout << " Options" << endl;
    cout << " 1. To open a tile" << endl;
    cout << " 2. To flag/unflag a tile" << endl;
    cout << " 3. To end the game" << endl;
    cout << endl;
    cout << endl;
    cout << " ▶ Remaining flags: " << remainingFlags + 1 << " 🏴";
    cout << endl;
    cout << endl;
    cout << "                    0   "<< "1   " << "2   " << "3   "<< "4   "<< "5   "<< "6   "<< "7   "<< "8   "<< endl;
    cout << "                    ▼   "<< "▼   " << "▼   " << "▼   "<< "▼   "<< "▼   "<< "▼   "<< "▼   "<< "▼   "<< endl;
    cout << endl;

    printDisplay();

    //Asks user for interaction
    cout << endl;
    cout << "Score:"<< score <<" point(s)" << endl;
    cout << "Type your option: ";
    cin >> option;
    cout << endl;
    return option;
}

int checkCoordinates(int x, int y){
  if(x < 0 || x > boardSize - 1){
      return 0;
  }
  else if(y < 0 || y > boardSize - 1){
      return 0;
  }
  return 1;

}

int gameOver() {
  cout << endl;
  cout << "Game Over!" << endl;
  cout << endl;
  cout << endl;

  return 0;
}

void askForCoordinates(int& xCoord, int& yCoord){
    cout <<"Please type Row and Column coordinates to open a tile." << endl;

    cin >> xCoord;//This is the Row (horizontal) coordinate
    cin >> yCoord;//This is the Column (vertical) coordinate

}

void getValidCoordinates(int& x, int& y){
    int coordinatesAreValid = 0;
    int attempt = 0;

    do{
        if(attempt > 0){
            cout << "Invalid Row or Column coordinates, please try again." << endl;
        }
        askForCoordinates(x, y);
        coordinatesAreValid = checkCoordinates(x, y);
        attempt++;

    }while(!coordinatesAreValid);
}

void setTileEmpty(int& x, int& y){
    display[y][x] = emptyTile;
}

void setTileFlagged(int& x, int& y){
    display[y][x] = flaggedTile;
}

void setTileCovered(int& x, int& y)
{
    display[y][x] = coveredTile;
}

int openEmptyTiles(int xCoord, int yCoord){

    if(!checkCoordinates(xCoord, yCoord)){//We check if the coordinates lead to a valid tile
        return 1;
    }
    if(display[yCoord][xCoord] != coveredTile){//We only want to keep going on covered tiles
      return 0;
    }

    else if(gameBoard[yCoord][xCoord] > 0){//If we found a number around a bomb, we must stop
        stringstream ss;

        ss << gameBoard[yCoord][xCoord];
        string str = ss.str();

        display[yCoord][xCoord] = str;

        remainingTiles--;
        return 0;
    }
    else{//We then open this one and proceed to the adjacent tiles
        setTileEmpty(xCoord, yCoord);
        remainingTiles--;

        if(yCoord > 0){
         openEmptyTiles(xCoord -1,yCoord -1);//Upper left corner
         openEmptyTiles(xCoord, yCoord -1);//Above
         openEmptyTiles(xCoord +1,yCoord -1);//Upper right corner
         }
        if(xCoord > 0){

        openEmptyTiles(xCoord -1,yCoord);//Left
        }
        if(xCoord < boardSize -1){
          openEmptyTiles(xCoord +1,yCoord);//Right
        }
        if(yCoord < boardSize - 1){
         openEmptyTiles(xCoord -1,yCoord +1);//Lower left corner
         openEmptyTiles(xCoord, yCoord +1);//Below
         openEmptyTiles(xCoord +1,yCoord +1);//Lower right corner
        }

    }

}

void displayAnswer(){
    for (int y = 0; y < boardSize; y++){
        for(int x = 0; x < boardSize; x++){

            if(display[y][x] == coveredTile || display[y][x] == flaggedTile){

                if(gameBoard[y][x] == -1){
                    display[y][x] = bombTile;
                }

                else if(gameBoard[y][x] == 0){
                    display[y][x] = emptyTile;
                }

                else{
                    stringstream ss;
                    ss << gameBoard[y][x];
                    string str = ss.str();
                    display[y][x] = str;
                }
            }
        }
    }
}

int openTile(){
    int xCoord;
    int yCoord;

    getValidCoordinates(xCoord, yCoord);

    if(gameBoard[yCoord][xCoord] == BOMB){
        cout << endl << "BOOOM! You have found a bomb ;)" << endl;
        return gameOver();
    }

    else if(display[yCoord][xCoord] == flaggedTile){
        cout <<"You cannot open flagged tiles!"<< endl;
        return 1;
    }

    else if(gameBoard[yCoord][xCoord] == EMPTY_TILE){
        openEmptyTiles(xCoord, yCoord);
    }

    else{
        stringstream ss;
        ss << gameBoard[yCoord][xCoord];
        string str = ss.str();
        if(display[yCoord][xCoord]!=str){
            display[yCoord][xCoord] = str;
            remainingTiles--;
        }
    }

    if( remainingTiles == 0){
          return 0;
      }

    return 1;
}

void flagTile(){
    int xCoord;
    int yCoord;
    getValidCoordinates(xCoord, yCoord);

    if(display[yCoord][xCoord] == flaggedTile){
        setTileCovered(xCoord, yCoord);
        remainingFlags++;
    }

    else if (remainingFlags >= 0) {
        setTileFlagged(xCoord, yCoord);
        remainingFlags--;
    }else {
        cout << "You cannot flag any more tiles!"<< endl;
        cout << endl;
    }
}

void setBombs(){
    int placedBombs = 0;
    srand(time(NULL));//We need to seed the randomizer

    while (placedBombs < bombs){//Generating and storing bomb coordinates
        int x = rand() % 8;
        int y = rand() % 8;

        if(gameBoard[y][x] == 0){
            gameBoard[y][x] = -1;//Places a bomb on the game board
            bombCoords[placedBombs][0] = y;//We then save the coordinates for that bomb
            bombCoords[placedBombs][1] = x;
            placedBombs++;
        }

    }
}

void setTips(){
    for(int c = 0; c < bombs; c++){//Filling tiles around the bombs with tips

        int yCoord = bombCoords[c][0];//Get the coordinates for the bombs
        int xCoord = bombCoords[c][1];

        if(yCoord > 0 && gameBoard[yCoord - 1][xCoord] > -1){//Add above bomb
            gameBoard[yCoord - 1][xCoord]++;//Adding to the tile adjacent to a bomb
        }

        if(yCoord > 0 && xCoord - 1 >= 0 && gameBoard[yCoord - 1][xCoord - 1] > -1){//Add upper left corner
            gameBoard[yCoord - 1][xCoord - 1]++;//Adding to the tile adjacent to a bomb
        }

        if(yCoord > 0 && xCoord + 1 < boardSize && gameBoard[yCoord - 1][xCoord + 1] > -1){//Add upper right corner
            gameBoard[yCoord - 1][xCoord + 1]++;//Adding to the tile adjacent to a bomb
        }

        if(yCoord < boardSize && gameBoard[yCoord + 1][xCoord] > -1){//Add below bomb
            gameBoard[yCoord + 1][xCoord]++;//Adding to the tile adjacent to a bomb

        }
        if(yCoord < boardSize && xCoord - 1 >= 0 && gameBoard[yCoord + 1][xCoord - 1] > -1){//Add lower left corner
            gameBoard[yCoord + 1][xCoord - 1]++;//Adding to the tile adjacent to a bomb
        }

        if(yCoord < boardSize && xCoord + 1 < boardSize && gameBoard[yCoord + 1][xCoord + 1] > -1){//Add lover right corner
            gameBoard[yCoord + 1][xCoord + 1]++;//Adding to the tile adjacent to a bomb
        }

        if(xCoord > 0 && gameBoard[yCoord][xCoord - 1] > -1){//Add to the left of the bomb
            gameBoard[yCoord][xCoord - 1]++;//Adding to the tile adjacent to a bomb
        }

        if(xCoord < boardSize && gameBoard[yCoord][xCoord + 1] > -1){//Add to the right of the bomb
            gameBoard[yCoord][xCoord + 1]++;//Adding to the tile adjacent to a bomb
        }
    }
}

void fillDisplay(){
    for(int h = 0; h <= boardSize - 1; h++){//Fills the display as covered tiles
        for(int j = 0; j <= boardSize - 1; j++){
                setTileCovered(h,j);
        }
    }

}

int main(){

    setBombs();
    setTips();
    fillDisplay();

//Game Execution Loop
    int gameEnded = 1;

    while (gameEnded != 0){

      int option = 0;

      option = showOptions();

      if(option == OPEN_TILE){
          gameEnded = openTile();
      }

      else if(option == FLAG_TILE){
          flagTile();
      }

      else if(option == EXIT_GAME){
          gameEnded = gameOver();
      }

      else {
          cout<< "Invalid option, please insert a valid option to proceed." << endl;
      }
    }

    if(remainingTiles == 0){

        cout <<"Congratulations, you won!"<< endl;
    }

    displayAnswer();
    printDisplay();

    return(0);
}

File: minesweeper.hs
import Data.List
import qualified Data.Set as Set
import Data.Char
import System.IO
import System.IO.Unsafe
import System.Random

--Game Tile Constants
coveredTile = '■'
emptyTile = '□'
flaggedTile = '🏴'
bombTile = '💣'

--Game Initial functions
getAmountBombs :: Int -> Float -> Int
getAmountBombs boardSize bombDensity = round ( (fromIntegral boardSize) * bombDensity * 10) :: Int

getInitialGameboard :: Int -> [[Int]]
getInitialGameboard boardSize = replicate boardSize (replicate boardSize 0)

getInitialDisplay :: Int -> [[Char]]
getInitialDisplay boardSize = replicate boardSize (replicate boardSize coveredTile)

--Printing related functions
printBoard :: [[Int]] -> IO()
printBoard gameBoard = do
	sequence_ (map print gameBoard) --Sequence will execute the IO() commands that are in the mapping

printDisplay :: [[Char]] -> IO()
printDisplay display = do
	let upperNums = formatNumGuide [0..((length display) - 1)]
	let formattedBoard = [("   " ++ upperNums)] ++ ["   " ++ (concat(replicate (length display) "  ▼"))] ++ formatBoard display

	sequence_ (map putStrLn formattedBoard)

	where
		formatNumGuide [] = " "
		formatNumGuide (n:ns)
			|n <= 9 = "  " ++ (intToDigit n) : formatNumGuide ns
			|otherwise = (intToDigit n) : formatNumGuide ns

formatBoard :: [String] -> [String]
formatBoard board =  formatBoardAux board (length board) 0

	where 
		formatBoardAux :: [String] -> Int -> Int -> [String]
		formatBoardAux lines 0 nLine = lines
		formatBoardAux (x:xs) len nLine
			|nLine <= 9 = ["  " ++ (intToDigit (nLine)) : "▶ "  ++ (formatLine x)] ++ formatBoardAux xs (length xs) (nLine + 1)
			|otherwise = [(intToDigit (nLine)) : "▶ " ++ (formatLine x)] ++ formatBoardAux xs (length xs) (nLine + 1)

formatLine :: String -> String
formatLine [] = []
formatLine (c:cs) = c : "  " ++ formatLine cs


--List operation related functions

editBoardAt :: [[a]] -> (Int,Int) -> a -> [[a]]
editBoardAt board coords element = editRowAt board coords 0 element

editRowAt :: [[a]] -> (Int,Int) -> Int -> a -> [[a]]
editRowAt [] coords nRow element = []
editRowAt (b:bs) (xCoord,yCoord) nRow element 
	|nRow == xCoord = (editLineAt b yCoord 0 element) : bs --change element in this line and put them back together
	|otherwise = b : (editRowAt bs (xCoord,yCoord) (nRow + 1) element) --Keep iterating

editLineAt :: [a] -> Int -> Int -> a -> [a]
editLineAt [] yCoord nColumn element = []
editLineAt (l:ls) yCoord nColumn element 
	|nColumn == yCoord = element : ls -- Change the element in that position
	|otherwise = l : (editLineAt ls yCoord (nColumn + 1) element)

getElement ::[[a]] -> (Int, Int) -> a
getElement matrix (x, y) = matrix !! x !! y

-- Method that return the adjacent positions of a coordinate
getAdjCoords :: Int -> (Int,Int)  -> [(Int,Int)]
getAdjCoords boardSize (x,y) = validateCoordinates [upperLeft x y, up x y, upperRight x y, left x y, right x y,
 lowerLeft x y, down x y, lowerRight x y] boardSize
	
	where 
		upperLeft x y = (x-1,y-1)

		up x y = (x-1,y)
		
		upperRight x y = (x-1,y+1)
		
		left x y = (x,y-1) 
		
		right x y = (x,y+1)
		
		lowerLeft x y = (x+1,y-1)
		
		down x y = (x+1,y)
		
		lowerRight x y = (x+1,y+1)


getHintCoords :: [(Int, Int)] -> [[Int]] -> [(Int, Int)] -- Get only 'hints' (element > 1) of the getAdjCoordsList
getHintCoords [] gBoard = []
getHintCoords (c:cs) gBoard = [(x,y) | (x,y) <- (getAdjCoords (length gBoard) c), getElement gBoard (x,y) > 0] ++ getHintCoords cs gBoard


-- Evaluates valid coordinates from a list
validateCoordinates :: [(Int,Int)] -> Int -> [(Int,Int)]
validateCoordinates [] boardSize = []
validateCoordinates ((xCoord,yCoord):xs) boardSize
    | (xCoord < boardSize && yCoord < boardSize) && (xCoord >= 0 && yCoord >= 0) = (xCoord,yCoord):validateCoordinates xs boardSize
    | otherwise = validateCoordinates xs boardSize

-- Returns a list of random coodinates given a threshold
getRandomCoords :: Int -> Int -> [(Int,Int)]
getRandomCoords amount boardSize = genRandomCoords amount boardSize []

--Generates a list of different coordinates (randomly choosen)
genRandomCoords :: Int -> Int -> [(Int,Int)] -> [(Int,Int)]
genRandomCoords 0 bSize coords = coords
genRandomCoords amount bSize [] = genRandomCoords (amount - 1) bSize [getRndTuple (bSize - 1)]
genRandomCoords amount bSize coords = do
	let sample = getRndTuple (bSize - 1)
	if (sample `elem` coords) == True then genRandomCoords amount bSize coords
		else genRandomCoords (amount - 1) bSize (sample : coords)

--Generates a tuple of random Int
getRndTuple :: Int -> (Int,Int)
getRndTuple threshold = (getRandomIntR(0,threshold), getRandomIntR(0,threshold))

--Generates a random Int from a range
getRandomIntR :: (Int,Int) -> Int
getRandomIntR (x,y) = unsafePerformIO (randomRIO (x,y))

--Gets all coordinates for adjacent empty spaces
getEmptyAdj :: [[Int]] -> (Int,Int) -> [(Int,Int)]
getEmptyAdj board point = getAdjIfEmpty board [point] [(x,y) | (x,y) <- (getAdjCoords (length board) point), isEmptyTile board (x,y)]

getAdjIfEmpty :: [[Int]] -> [(Int,Int)] -> [(Int,Int)] -> [(Int, Int)]
getAdjIfEmpty board results [] = results
getAdjIfEmpty board results toCheck = do
	--This list comprehension here gets all points adjacent to the coordinates in toCheck, if they are emptyTiles
	let newResults = results ++ [(x,y) | (x,y) <- (toCheck), isEmptyTile board (x,y), not ((x,y) `elem` results)]

	--This list comprehension here gets all adjacent emptyTiles from toCheck, if they are not yet in the newResults
	--'concat' does [[a]] -> [a] and 'delDups' removes all duplicates from the list
	let newToCheck = delDups ([(x,y) | (x,y) <- (concat (map (getAdjCoords (length board)) toCheck)), isEmptyTile board (x,y), not((x,y) `elem` newResults)])

	--RecursiveStep
	getAdjIfEmpty board newResults newToCheck

--Checks if coordinate leads to an EmptyTile
isEmptyTile :: [[Int]] -> (Int,Int) -> Bool
isEmptyTile board point = (getElement board point) == 0

coordsAreValid :: [(Int,Int)] -> Int-> Bool
coordsAreValid coords boardSize = (length (validateCoordinates coords boardSize)) > 0

-- Game title
getTitle :: String
getTitle = "\n *** Minesweeper *** \n"

-- Menu Options
getMenu :: String
getMenu  = "Options\n  1. Open Tile\n  2. Flag/Unflag Tile\n  3. Exit Game\n\n Choose your Option: "

-- Asking for coordinates
askForCoordsOpen :: String
askForCoordsOpen = "Please type Row and Column coordinates to open a tile: "

askForCoordsFlag :: String
askForCoordsFlag = "Please type Row and Column coordinates to Flag/Unflag a tile: "

gameover :: Int -> [[Int]] -> [[Char]] -> IO()
gameover points gBoard display = do
	let revealedDisplay = openAllTiles gBoard display
	printDisplay revealedDisplay
	putStrLn ("\nGame Over! You achieved " ++ (show points) ++ " point(s)!")

getCoordFromList :: [Int] -> (Int,Int)
getCoordFromList list = ((list !! 0), (list !! 1))

openTiles :: [(Int, Int)] -> [[Int]] -> [[Char]]  -> [[Char]]
openTiles [] board display = display
openTiles (c:cs) board display
	| getElement display c /= coveredTile = openTiles cs board display
	| getElement board c == 0 = openTiles cs board (editBoardAt display c emptyTile)
	| getElement board c > 0 = openTiles cs board (editBoardAt display c (intToDigit (getElement board c)))
	| getElement board c == -1 = openTiles cs board (editBoardAt display c bombTile)

openAllTiles :: [[Int]] -> [[Char]]  -> [[Char]]
openAllTiles board display = openTiles [(x, y) | x <- [0.. (length board)-1], y <- [0..(length board)-1]] board display
	
insertHints :: [[Int]] -> [(Int,Int)] ->[[Int]]
--The list comprehension here evaluates to all coordinates around the bombs (excluding bombs themselves)
insertHints board bombCoords = insertRec board [(x,y) | (x,y) <- (concat(map (getAdjCoords (length board)) bombCoords)),
 not ((getElement board (x,y)) == -1)]

	where 
		insertRec board [] = board
		insertRec board (c:cs) = insertRec (editBoardAt board c ((getElement board c) + 1)) cs

insertBombs :: [[Int]] -> [(Int,Int)] -> [[Int]]
insertBombs gBoard [] = gBoard
insertBombs gBoard (c:cs) = insertBombs (editBoardAt gBoard c (-1)) cs

delDups :: Ord a => [a] -> [a]
delDups = Set.toList . Set.fromList


start :: [[Int]] -> [[Char]] -> Int -> Int -> Int-> IO()
start gBoard display flags points tilesLeft = do

	if tilesLeft == 0 then gameover points gBoard display

	else do
		putStrLn getTitle
		
		putStrLn (" Points: " ++ show points)
		putStrLn (" Remaining Flags: " ++ show flags ++ flaggedTile : "\n")	
	

		printDisplay display	

		putStrLn ("\n" ++ getMenu)	

		option' <- getLine	

		let option = read option' :: Int	

		if option == 1 then do --Open a tile
				putStrLn askForCoordsOpen
				input <- getLine	

				let coords = getCoordFromList (map read $ words input :: [Int])	

				if coordsAreValid [coords] (length gBoard) && (getElement display coords) /= flaggedTile then --If the coords are valid and are not flagged
					if getElement gBoard coords == -1 then gameover points gBoard display -- If the coords lead to a bomb, it's game over
					else do --In this case the coords lead to an empty space or a hint tile	

						let coordsToOpen = coords : getEmptyAdj gBoard coords
						let hints = delDups (getHintCoords coordsToOpen gBoard)
						let newScore = length (coordsToOpen)
						let remainingTiles = length (coordsToOpen ++ hints)	

						start gBoard (openTiles (coordsToOpen ++ hints) gBoard display) flags (points + newScore) remainingTiles	

				else do --Invalid/Flagged coordinates
					putStrLn "Invalid coordinates! The tile is flagged or could not be found."
					start gBoard display flags points tilesLeft	
	

		else if option == 2 then do --Flag/Unflag a tile	

			putStrLn askForCoordsFlag
			input <- getLine	

			let coords = getCoordFromList (map read $ words input :: [Int])	
	

			if flags > 0 && coordsAreValid [coords] (length gBoard) && getElement display coords == coveredTile then --Flags are available, coordinates are valid and lead to a covered tile
				start gBoard (editBoardAt display coords flaggedTile) (flags - 1) points (tilesLeft - 1)	

			else if coordsAreValid [coords] (length gBoard) && getElement display coords == flaggedTile then --Unflag a tile
				start gBoard (editBoardAt display coords coveredTile) (flags + 1) points (tilesLeft + 1)	

			else start gBoard display flags points tilesLeft --Nothing happens, proceed with the game	
	

		else if option == 3 then gameover points gBoard display	

		else do --Not a valid option
			putStrLn "That was not a valid option!"
			start gBoard display flags points tilesLeft


main = do
	
	--Game Constants
	let boardSize = 9 -- 9x9 board
	let bombDensity = 0.1 -- 10% of the game board

	let amountBombs = getAmountBombs boardSize bombDensity
	let amountFlags = amountBombs --A flag for each bomb
	let iPoints = 0

	let bombCoords = getRandomCoords amountBombs (boardSize)

	let iGameboard = insertHints (insertBombs (getInitialGameboard boardSize) bombCoords) bombCoords --Insert bombs then their hints in the board
	let iDisplay = getInitialDisplay boardSize

	let iRemainingTiles = (boardSize * boardSize) - amountBombs
	

	start iGameboard iDisplay amountFlags iPoints iRemainingTiles
